File: Design-Patterns-Using-CSharp/Template/TemplatePatternPerson.cs
using System;

abstract class AbstractWorker
{
    // The template method
    public void DailyRoutine()
    {
        WakeUp();
        EatBreakfast();
        GoToWork();
        Work();
        ComeBackHome();
        EatDinner();
        Sleep();
    }

    protected void WakeUp()
    {
        Console.WriteLine("Wake Up");
    }

    protected void EatBreakfast()
    {
        Console.WriteLine("Eat Breakfast");
    }

    protected void GoToWork()
    {
        Console.WriteLine("Go to work");
    }

    protected abstract void Work();

    protected void ComeBackHome()
    {
        Console.WriteLine("Come back Home");
    }

    protected void EatDinner()
    {
        Console.WriteLine("Eat dinner");
    }

    protected void Sleep()
    {
        Console.WriteLine("Sleep");
    }
}

class Doctor : AbstractWorker
{
    protected override void Work()
    {
        Console.WriteLine("...Treat Patients...");
    }
}

class FireFighter : AbstractWorker
{
    protected override void Work()
    {
        Console.WriteLine("...Fight Fire...");
    }
}

class SuperHero : AbstractWorker
{
    protected override void Work()
    {
        Console.WriteLine("...Save the world!...");
    }
}

// Client code
class TemplatePatternPerson
{
    public static void Main(string[] args)
    {
        Doctor doctor = new Doctor();
        doctor.DailyRoutine();
        Console.WriteLine();

        FireFighter fireFighter = new FireFighter();
        fireFighter.DailyRoutine();
        Console.WriteLine();

        SuperHero superHero = new SuperHero();
        superHero.DailyRoutine();
    }
}

File: Design-Patterns-Using-Kotlin/Template/TemplatePatternPerson.kt
abstract class AbstractWorker {
    fun dailyRoutine() {
        wakeUp()
        eatBreakfast()
        goToWork()
        work()
        comeBackHome()
        eatDinner()
        sleep()
    }

    fun wakeUp() {
        println("Wake Up")
    }

    fun eatBreakfast() {
        println("Eat Breakfast")
    }

    fun goToWork() {
        println("Go to work")
    }

    abstract fun work()

    fun comeBackHome() {
        println("Come back Home")
    }

    fun eatDinner() {
        println("Eat dinner")
    }

    fun sleep() {
        println("Sleep")
    }
}

class Doctor : AbstractWorker() {
    override fun work() {
        println("...Treat Patients...")
    }
}

class FireFighter : AbstractWorker() {
    override fun work() {
        println("...Fight Fire...")
    }
}

class SuperHero : AbstractWorker() {
    override fun work() {
        println("...Save the world!...")
    }
}


// Client code
fun main() {
    val doctor = Doctor()
    doctor.dailyRoutine()
    println()

    val fireFighter = FireFighter()
    fireFighter.dailyRoutine()
    println()

    val superHero = SuperHero()
    superHero.dailyRoutine()
}

/*
Wake Up
Eat Breakfast
Go to work
...Treat Patients...
Come back Home
Eat dinner
Sleep

Wake Up
Eat Breakfast
Go to work
...Fight Fire...
Come back Home
Eat dinner
Sleep

Wake Up
Eat Breakfast
Go to work
...Save the world!...
Come back Home
Eat dinner
Sleep
*/